src/dns.rs

```diff
@@ -0,0 +1,510 @@
+        let name = host
+            .parse::<Name>()
+            .map_err(|error| DnsError::InvalidHost(host.to_owned(), error))?;
+
+        self.resolve_name(name).map_err(Into::into)
+    }
+
+    /// Should be only called from outside tokio. From inside tokio, Async version should be used.
+    fn resolve_name(&self, name: Name) -> Result<IntoIter<IpAddr>, io::Error> {
+        let future = self.construct_future(name);
+        // Will panic if called from inside tokio.
+        tokio::runtime::current_thread::block_on_all(future)
+    }
+
+    fn construct_future(&self, name: Name) -> ResolveFuture {
+        // try to parse as a regular IpAddr first
+        name.as_str()
+            .parse::<IpAddr>()
+            .ok()
+            .map(ResolveFuture::IpAddr)
+            // Else start resolution
+            .unwrap_or_else(|| match self.resolver.as_ref() {
+                Some(resolver) => ResolveFuture::Custom(resolver.lookup_ip(name.as_str()), name),
+                None => ResolveFuture::System(name),
+            })
+    }
+}
+
+impl Resolve for DnsResolver {
+    type Addrs = IntoIter<IpAddr>;
+    /// A Future of the resolved set of addresses.
+    type Future = ResolveFuture;
+    /// Resolve a hostname.
+    fn resolve(&self, name: Name) -> Self::Future {
+        self.construct_future(name)
+    }
+}
+
+pub enum ResolveFuture {
+    /// Resolved
+    IpAddr(IpAddr),
+    /// Resolving using custom DNS servers
+    Custom(BackgroundLookupIp, Name),
+    /// Resolve using system resolver servers
+    System(Name),
+}
+
+impl ResolveFuture {
+    fn system_resolve(name: &Name) -> Poll<IntoIter<IpAddr>, io::Error> {
+        match (name.as_str(), 0).to_socket_addrs() {
+            Ok(ips) => Poll::Ok(Async::Ready(
+                ips.map(|socket_addr| socket_addr.ip())
+                    .collect::<Vec<_>>()
+                    .into_iter(),
+            )),
+            Err(error) => Poll::Err(error),
+        }
+    }
+}
+
+impl Future for ResolveFuture {
+    type Item = IntoIter<IpAddr>;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Self::Item, Self::Error> {
+        match self {
+            ResolveFuture::IpAddr(addr) => Poll::Ok(Async::Ready(vec![addr.clone()].into_iter())),
+            ResolveFuture::Custom(future, name) => match future.poll() {
+                Poll::Ok(Async::NotReady) => Poll::Ok(Async::NotReady),
+                Poll::Ok(Async::Ready(ips)) => {
+                    Poll::Ok(Async::Ready(ips.iter().collect::<Vec<_>>().into_iter()))
+                }
+                Poll::Err(error) => {
+                    match error.kind() {
+                        ResolveErrorKind::NoRecordsFound { .. } => debug!("No records found"),
+                        _ => {
+                            error!(message = "Error while resolving Domain Name", error = ?error);
+                        }
+                    }
+                    ResolveFuture::system_resolve(name)
```

Comment:
This is blocking so we should use a https://docs.rs/tokio-threadpool/0.1.16/tokio_threadpool/fn.blocking.html call.

Reply:
:man_facepalming: True